lib/tbot/service_spiffe_workload_api.go

```diff
@@ -0,0 +1,489 @@
+		s.botCfg.CertificateTTL,
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Convert the private key to PKCS#8 format as per SPIFFE spec.
+	pkcs8PrivateKey, err := x509.MarshalPKCS8PrivateKey(privateKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Convert responses from the Teleport API to the SPIFFE Workload API
+	// format.
+	svids := make([]*workloadpb.X509SVID, len(res.Svids))
+	for i, svidRes := range res.Svids {
+		svids[i] = &workloadpb.X509SVID{
+			// Required. The SPIFFE ID of the SVID in this entry
+			SpiffeId: svidRes.SpiffeId,
+			// Required. ASN.1 DER encoded certificate chain. MAY include
+			// intermediates, the leaf certificate (or SVID itself) MUST come first.
+			X509Svid: svidRes.Certificate,
+			// Required. ASN.1 DER encoded PKCS#8 private key. MUST be unencrypted.
+			X509SvidKey: pkcs8PrivateKey,
+			// Required. ASN.1 DER encoded X.509 bundle for the trust domain.
+			Bundle: trustBundle,
+			Hint:   svidRes.Hint,
+		}
+		cert, err := x509.ParseCertificate(svidRes.Certificate)
+		if err != nil {
+			return nil, trace.Wrap(err, "parsing issued svid received from server")
+		}
+		// Log a message which correlates with the audit log entry and can
+		// provide additional metadata about the client.
+		s.log.WithFields(logrus.Fields{
+			"svid_type":     "x509",
+			"spiffe_id":     svidRes.SpiffeId,
+			"serial_number": serialString(cert.SerialNumber),
+			"hint":          svidRes.Hint,
+			"not_after":     cert.NotAfter,
+			"not_before":    cert.NotBefore,
+			"dns_sans":      cert.DNSNames,
+			"ip_sans":       cert.IPAddresses,
+			"serial":        cert.SerialNumber,
+		}).Info("Issued SVID for workload")
+	}
+
+	return svids, nil
+}
+
+// FetchX509SVID generates and returns the X.509 SVIDs available to a workload.
+// It is a streaming RPC, and sends renewed SVIDs to the client before they
+// expire.
+// Implements the SPIFFE Workload API FetchX509SVID method.
+func (s *SPIFFEWorkloadAPIService) FetchX509SVID(
+	_ *workloadpb.X509SVIDRequest,
+	srv workloadpb.SpiffeWorkloadAPI_FetchX509SVIDServer,
+) error {
+	renewCh, unsubscribe := s.trustBundleBroadcast.subscribe()
+	defer unsubscribe()
+	s.log.Info("FetchX509SVID stream opened by workload")
+	defer s.log.Info("FetchX509SVID stream has closed")
+
+	for {
+		s.log.Info("Starting to issue X509 SVIDs to workload")
+
+		svids, err := s.fetchX509SVIDs(srv.Context())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		err = srv.Send(&workloadpb.X509SVIDResponse{
+			Svids: svids,
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		s.log.Debug(
+			"Finished issuing SVIDs to workload. Waiting for next renewal interval or CA rotation",
+		)
+
+		select {
+		case <-srv.Context().Done():
+			s.log.Debug("Context closed, stopping SVID stream")
+			return nil
+		case <-time.After(s.botCfg.RenewalInterval):
+			s.log.Debug("Renewal interval reached, renewing SVIDs")
+			// Time to renew the certificate
+			continue
```

Comment:
Elsewhere we wait for the channel associated with the `rootReloadBroadcaster` for deciding that we're undergoing a cert renewal, should we do the same here (or is this something else / I'm just confused)?

https://github.com/gravitational/teleport/blob/fe345d491bcc2378493dbe1156e8363ee41304e3/lib/tbot/service_spiffe_workload_api.go#L273-L279

Reply:
`reloadCh` is fulfilling the same role here. Essentially, we have another goroutine which watches `rootReloadBroadcaster` and then fetches the new CA bundle and then sends a message on `reloadCh`. This just allows us to do a little bit of work before letting the RPCs know and prevents all the RPCs having to complete that work individually.